### PR TITLE
feat(commerce): Add method for generating ISBN-10 and ISBN-13

### LINKED ISSUE #2196: ISBN 10 and ISBN13
### Clear and concise description of the problem

Add ISBN 10 and 13 generators to Faker. They look like this:

```
isbn10: "0140449337"
isbn13: "978-0140449334"
```

### Suggested solution

Add it to `commerce` or add a new root level generator for books of which this ISBN generator would be a member of.

### Alternative

_No response_

### Additional context

This is already implemented in faker libraries in other ecosystems: https://faker.readthedocs.io/en/master/providers/faker.providers.isbn.html

### PR DESCRIPTION
**Pull request by RobinvanderVliet**
This is my first pull request contributing some actual code to this project, so any feedback is welcome!

This merge request resolves issue #2196.

### CODE REVIEW

File: src/modules/commerce/index.ts [hunk 1]
```diff
@@ -266,4 +338,82 @@ export class CommerceModule {
...
+        } = {}
+  ): string {
+    if (typeof options === 'number') {
+      options = { variant: options };
+    }
+
+    const { variant = 13, separator = '-' } = options;
+
+    const prefix = '978';
+    const group: string = this.faker.number.int({ min: 0, max: 1 }).toString();
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]: You might want to use `faker.helpers.objectKey(ISBN_LENGTH_RULES)` here instead.

File: src/modules/commerce/index.ts [hunk 2]
```diff
@@ -266,4 +338,82 @@ export class CommerceModule {
...
+    const data: string[] = [prefix, group, registrant, publication];
+    if (variant === 10) {
+      data.shift();
+    }
+
+    const isbn: string = data.join('');
+
+    let checksum = 0;
+    for (let i = 0; i < variant - 1; i++) {
+      const weight: number = variant === 10 ? i + 1 : i % 2 ? 3 : 1;
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]: IMO in cases like these it is obvious that weight is a number, you don't need the explicit type definition.

File: test/commerce.spec.ts [hunk 3]
```diff
@@ -158,6 +169,55 @@ describe('commerce', () => {
           );
         });
       });
+
+      describe(`isbn()`, () => {
+        it('should return ISBN-13 with hyphen separators when not passing arguments', () => {
```
Comment (ST-DDT) [inline test/commerce.spec.ts]: Please add a test for isn10+13 each that validates the generated isbn with validatorjs's [`isISBN()`](https://github.com/validatorjs/validator.js/blob/master/src/lib/isISBN.js)

See here for example: https://github.com/faker-js/faker/blob/4a15bd0d4eea5e5f6a7d1e6be0c2bf9e59feb069/test/internet.spec.ts#L555

Please pass the variant as options to the check, so it checks for the exact variant only.

**Review by ST-DDT — COMMENTED** [review]
Please note that the associated issue is marked as waiting for user interest, so this PR might not be merged immediately.

File: src/modules/commerce/index.ts (hunk 2 above)
Comment (RobinvanderVliet) [reply src/modules/commerce/index.ts]: I removed the explicit type definition. I am used to always adding them.

File: test/commerce.spec.ts (hunk 3 above)
Comment (RobinvanderVliet) [reply test/commerce.spec.ts]: I added the check to the unit tests.

File: src/modules/commerce/index.ts (hunk 1 above)
Comment (RobinvanderVliet) [reply src/modules/commerce/index.ts]: Done!

File: src/modules/commerce/index.ts [hunk 4]
```diff
@@ -1,6 +1,78 @@
 import type { Faker } from '../../faker';
 import { deprecated } from '../../internal/deprecated';
 
+// Source for official prefixes: https://www.isbn-international.org/range_file_generation
+const ISBN_LENGTH_RULES: {
+  [group: string]: Array<[rangeMaximum: number, length: number]>;
+} = {
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]: If we simplify the type here to:

```suggestion
const ISBN_LENGTH_RULES: Record<
  string,
  Array<[rangeMaximum: number, length: number]>
> = {
```

File: src/modules/commerce/index.ts [hunk 5]
```diff
@@ -266,4 +338,84 @@ export class CommerceModule {
...
+    if (typeof options === 'number') {
+      options = { variant: options };
+    }
+
+    const { variant = 13, separator = '-' } = options;
+
+    const prefix = '978';
+    const group: string = this.faker.helpers
+      .objectKey(ISBN_LENGTH_RULES)
+      .toString();
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]: Then we can simplify this to:

```suggestion
    const [group, groupRules] =
      this.faker.helpers.objectEntry(ISBN_LENGTH_RULES);
```

File: src/modules/commerce/index.ts [hunk 6]
```diff
@@ -266,4 +338,84 @@ export class CommerceModule {
...
+    const element: string = this.faker.string.numeric(8);
+
+    const registrantLength: number = ISBN_LENGTH_RULES[group].find(
+      ([rangeMaximum]): boolean =>
+        parseInt(element.slice(0, -1)) <= rangeMaximum
+    )[1];
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]: Then you can change `ISBN_LENGTH_RULES[group]` to `groupRules`. Also the `parseInt(element.slice(0, -1))` appears to be static, so IMO you can calculate that only once and reuse the value in the filter.

```suggestion
    const elementValue = parseInt(element.slice(0, -1));

    const registrantLength = groupRules.find(
      ([rangeMaximum]) => elementValue <= rangeMaximum
    )[1];
```

File: src/modules/commerce/index.ts [hunk 7]
```diff
@@ -266,4 +338,84 @@ export class CommerceModule {
...
+      options = { variant: options };
+    }
+
+    const { variant = 13, separator = '-' } = options;
+
+    const prefix = '978';
+    const group: string = this.faker.helpers
+      .objectKey(ISBN_LENGTH_RULES)
+      .toString();
+    const element: string = this.faker.string.numeric(8);
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]: IMO there is no need to explicitly define that this is a string here.

```suggestion
    const element = this.faker.string.numeric(8);
```

File: src/modules/commerce/index.ts [hunk 8]
```diff
@@ -266,4 +338,84 @@ export class CommerceModule {
...
+
+    const registrantLength: number = ISBN_LENGTH_RULES[group].find(
+      ([rangeMaximum]): boolean =>
+        parseInt(element.slice(0, -1)) <= rangeMaximum
+    )[1];
+
+    const registrant: string = element.slice(0, registrantLength);
+    const publication: string = element.slice(registrantLength);
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]: Same here

```suggestion
    const registrant = element.slice(0, registrantLength);
    const publication = element.slice(registrantLength);
```

File: src/modules/commerce/index.ts [hunk 9]
```diff
@@ -266,4 +338,84 @@ export class CommerceModule {
...
+
+    const data: string[] = [prefix, group, registrant, publication];
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]:
```suggestion
    const data = [prefix, group, registrant, publication];
```

File: src/modules/commerce/index.ts [hunk 10]
```diff
@@ -266,4 +338,84 @@ export class CommerceModule {
...
+    if (variant === 10) {
+      data.shift();
+    }
+
+    const isbn: string = data.join('');
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]:
```suggestion
    const isbn = data.join('');
```

Comment (RobinvanderVliet) [reply src/modules/commerce/index.ts]: Thank you for your review! I've processed your comments. :+1:

File: src/modules/commerce/index.ts [hunk 11]
```diff
@@ -258,4 +331,83 @@ export class CommerceModule {
...
+   *
+   * @example
+   * faker.commerce.isbn() // '978-0-692-82459-7'
+   * faker.commerce.isbn(10) // '1-155-36404-X'
+   * faker.commerce.isbn(13) // '978-1-60808-867-6'
+   * faker.commerce.isbn({ separator: ' ' }) // '978 0 452 81498 1'
+   * faker.commerce.isbn({ variant: 10, separator: ' ' }) // '0 940319 49 7'
+   * faker.commerce.isbn({ variant: 13, separator: ' ' }) // '978 1 6618 9122 0'
+   *
+   * @since 8.0.0
```
Comment (ST-DDT) [inline src/modules/commerce/index.ts]:
```suggestion
   * @since 8.1.0
```

**Review by ST-DDT — DISMISSED** [review]

**Review by ST-DDT — DISMISSED** [review]

**Comment by xDivisionByZerox** [comment]
**Team decision**

We accept this feature request. 

---

@RobinvanderVliet Please resolve to merge conflicts so we can open it for review.

**Comment by ST-DDT** [comment]
@RobinvanderVliet Could you please fix the merge conflicts?

**Comment by RobinvanderVliet** [comment]
I resolved the merge conflicts!

**Review by ST-DDT — APPROVED** [review]

**Review by xDivisionByZerox — APPROVED** [review]
I’m considering the use of a strategy pattern to improve the maintainability of the changed code. However, I have decided to accept the current implementation.

**Review by ejcheng — APPROVED** [review]

**Pull request merged**